### PR TITLE
[Core] Time discretization general BDF

### PR DESCRIPTION
**Pull request by rubenzorrilla**
While starting the migration from the old BDF coefficients process to the new time integration utilities I came up with the idea of having a sort of wrapping class for all the BDF classes. The idea is that the user sets the BDF order and the coefficients are accordingly computed.

I did it in such a way that both the "wrapping" BDF class (I called it `BDF`) has the same API as the `BDF1`, `BDF2`, etc ones. Two things to note:
- The specific order classes derive from the `BDF` new one (I required this to set a pointer to them)
- Now the return type is not a bounded array but an `std::vector<double>` (the number of BDF coefficients is not known a priori).

@philbucher What do you think about?

### CODE REVIEW

File: kratos/utilities/time_discretization.h [hunk 1]
```diff
@@ -25,47 +26,124 @@ class ProcessInfo; // forward-declaring to not having to include it here
...
+     * @brief Get the Time Order object
+     * Auxiliary method to get the order of the BDF scheme
+     * @return const unsigned int Order of the BDF scheme
+     */
+    const unsigned int GetTimeOrder() const;
+
+private:
+
+    const unsigned int mTimeOrder = 0; // Time order of the auxiliary BDF class
+    unique_ptr<BDF> mpAuxBDF = nullptr; // Pointer to an auxiliary BDF class with order
```
Comment (philbucher) [inline kratos/utilities/time_discretization.h]:
```suggestion
    Kratos::unique_ptr<BDF> mpAuxBDF = nullptr; // Pointer to an auxiliary BDF class with order
```

File: kratos/utilities/time_discretization.h [hunk 2]
```diff
@@ -25,47 +26,124 @@ class ProcessInfo; // forward-declaring to not having to include it here
...
+    /**
+     * @brief Get the Time Order object
+     * Auxiliary method to get the order of the BDF scheme
+     * @return const unsigned int Order of the BDF scheme
+     */
+    const unsigned int GetTimeOrder() const;
+
+private:
+
+    const unsigned int mTimeOrder = 0; // Time order of the auxiliary BDF class
```
Comment (philbucher) [inline kratos/utilities/time_discretization.h]:
```suggestion
    const std::size_t mTimeOrder = 0; // Time order of the auxiliary BDF class
```

File: kratos/utilities/time_discretization.h [hunk 3]
```diff
@@ -142,6 +220,7 @@ class GeneralizedAlpha
     double mNewmarkGamma=0.5;
 };
 
+inline std::size_t GetMinimumBufferSize(const BDF &rTimeDiscr) { return (rTimeDiscr.GetTimeOrder() + 1); }
```
Comment (philbucher) [inline kratos/utilities/time_discretization.h]:
```suggestion
inline std::size_t GetMinimumBufferSize(const BDF& rTimeDiscr) { return (rTimeDiscr.GetTimeOrder() + 1); }
```
to be consistent

File: kratos/utilities/time_discretization.h [hunk 4]
```diff
@@ -25,47 +26,124 @@ class ProcessInfo; // forward-declaring to not having to include it here
...
+    virtual std::vector<double> ComputeBDFCoefficients(double DeltaTime, double PreviousDeltaTime) const;
+
+    /**
+     * @brief Return the BDF coefficients
+     * This method computes the BDF coefficients for the time step
+     * stored in the variable DELTA_TIME of a ProcessInfo container
+     * @param rProcessInfo ProcessInfo container with DELTA_TIME
+     * @return std::vector<double> Vector containing the computed BDF coefficients
+     */
+    virtual std::vector<double> ComputeBDFCoefficients(const ProcessInfo &rProcessInfo) const;
```
Comment (philbucher) [inline kratos/utilities/time_discretization.h]:
```suggestion
    virtual std::vector<double> ComputeBDFCoefficients(const ProcessInfo& rProcessInfo) const;
```
most classes in the core do it like this, e.g. `Element`

**Review by philbucher — COMMENTED** [review]
don't forget to change the include of `<array>` to `<vector>`

**Comment by philbucher** [comment]
sorry I submitted the review too early, there is more coming :)

File: kratos/utilities/time_discretization.cpp [hunk 5]
```diff
@@ -25,35 +26,90 @@
 namespace Kratos {
 namespace TimeDiscretization {
 
-std::array<double, 2> BDF1::ComputeBDFCoefficients(const double DeltaTime) const
+void BDF::SetAuxBDFPointer(
+    const unsigned int TimeOrder,
+    unique_ptr<BDF> &rpAuxBDF)
```
Comment (philbucher) [inline kratos/utilities/time_discretization.cpp]:
```suggestion
    Kratos::unique_ptr<BDF> &rpAuxBDF)
```

File: kratos/utilities/time_discretization.cpp [hunk 6]
```diff
@@ -25,35 +26,90 @@
 namespace Kratos {
 namespace TimeDiscretization {
 
-std::array<double, 2> BDF1::ComputeBDFCoefficients(const double DeltaTime) const
+void BDF::SetAuxBDFPointer(
+    const unsigned int TimeOrder,
```
Comment (philbucher) [inline kratos/utilities/time_discretization.cpp]:
```suggestion
    const std::size_t TimeOrder,
```

File: kratos/utilities/time_discretization.h [hunk 7]
```diff
@@ -25,47 +26,124 @@ class ProcessInfo; // forward-declaring to not having to include it here
...
+    virtual ~BDF(){};
+
+    /**
+     * @brief Set the Aux BDF Pointer object
+     * This method sets a pointer to one of the auxiliary BDF
+     * classes according to the user defined time order
+     * @param TimeOrder BDF scheme time order
+     * @param rpAuxBDF Pointer to the auxiliary BDF scheme
+     */
+    static void SetAuxBDFPointer(
```
Comment (philbucher) [inline kratos/utilities/time_discretization.h]: why is this function needed?
The object was designed lightweight, and this gives the user quite some influence for manipulation

File: kratos/utilities/time_discretization.h [hunk 8]
```diff
@@ -25,47 +26,124 @@ class ProcessInfo; // forward-declaring to not having to include it here
...
-    std::array<double, 2> ComputeBDFCoefficients(double DeltaTime) const;
-    std::array<double, 2> ComputeBDFCoefficients(const ProcessInfo& rProcessInfo) const;
+
+    KRATOS_CLASS_POINTER_DEFINITION(BDF);
+
+    /**
+     * @brief Construct a new BDF object
+     * Auxiliary constructor for derived classes
+     */
+    BDF(){};
```
Comment (philbucher) [inline kratos/utilities/time_discretization.h]: I would not allow this, the user should always give a time-order, since this is a "wrapper"

**Review by philbucher — COMMENTED** [review]
Overall I am fine with your changes.
I made some comments regarding the usage, since this was designed to be lightweight

Also please check the tests in the MeshMovingApp, I am using it there

File: kratos/utilities/time_discretization.h (hunk 8 above)
Comment (rubenzorrilla) [reply kratos/utilities/time_discretization.h]: This is needed for derived classes. We can move it to a `protected` section.

File: kratos/utilities/time_discretization.h (hunk 7 above)
Comment (rubenzorrilla) [reply kratos/utilities/time_discretization.h]: It is needed in the constructor to select among the BDF1, BDF2, BDF3, etc. The nice point is that if you construct the "old" BDF objects the overhead is minimum. However, the new generic BDF gives more flexibility.

**Comment by rubenzorrilla** [comment]
> Overall I am fine with your changes.
> I made some comments regarding the usage, since this was designed to be lightweight
> 
> Also please check the tests in the MeshMovingApp, I am using it there

I ran the MeshMovingApp and FluidDynamicsApp tests and everything seems to be OK. Note that the MeshMovingApp is using your "old" implementation.

File: kratos/utilities/time_discretization.h (hunk 7 above)
Comment (rubenzorrilla) [reply kratos/utilities/time_discretization.h]: Furthermore, I think we should move it to the `private` section.

File: kratos/utilities/time_discretization.h (hunk 8 above)
Comment (rubenzorrilla) [reply kratos/utilities/time_discretization.h]: Moved to a `protected`section.

File: kratos/utilities/time_discretization.h (hunk 7 above)
Comment (rubenzorrilla) [reply kratos/utilities/time_discretization.h]: Moved to a `private` section.

**Comment by philbucher** [comment]
I will review again after it compiles :)

**Comment by rubenzorrilla** [comment]
> I will review again after it compiles :)

There's something in Travis. Let me check.

**Comment by rubenzorrilla** [comment]
@philbucher it was some missing `overrides`. I think it is ready to be re-reviewed.

File: kratos/utilities/time_discretization.h [hunk 9]
```diff
@@ -25,47 +26,126 @@ class ProcessInfo; // forward-declaring to not having to include it here
...
+    BDF(const unsigned int TimeOrder) : mTimeOrder(TimeOrder)
+    {
+        this->SetAuxBDFPointer(TimeOrder, mpAuxBDF);
+    };
+
+    /**
+     * @brief Destroy the BDF object
+     * Destructor of the BDF class
+     */
+    virtual ~BDF(){};
```
Comment (philbucher) [inline kratos/utilities/time_discretization.h]:
```suggestion
    virtual ~BDF() = default;
```

**Review by philbucher — COMMENTED** [review]
minor change, otherwise nice work :+1: 
=> I think you have to make it ready for review

**Review by philbucher — APPROVED** [review]

**Pull request merged**